indexer/cmd/indexer/cli.go

```diff
@@ -76,16 +79,44 @@ func runApi(ctx *cli.Context) error {
 	return api.Listen(apiCtx, cfg.API.Port)
 }
 
-func runAll(ctx *cli.Context) error {
-	// Run the indexer
-	go func() {
-		if err := runIndexer(ctx); err != nil {
-			log.NewLogger(log.ReadCLIConfig(ctx)).Error("Error running the indexer", "err", err)
+func runAll(cliCtx *cli.Context) error {
+	logger := log.NewLogger(log.ReadCLIConfig(cliCtx))
+
+	var wg sync.WaitGroup
+	errCh := make(chan error, 2) // represents that 2 goroutines will be running
+
+	_, sharedCancel := context.WithCancel(context.Background())
+	defer sharedCancel()
+
+	run := func(startFunc func(*cli.Context) error) {
+		wg.Add(1)
+		defer func() {
+			if err := recover(); err != nil {
+				log.NewLogger(log.ReadCLIConfig(cliCtx)).Error("halting on panic", "err", err)
+				debug.PrintStack()
+				errCh <- fmt.Errorf("panic: %v", err)
+			}
+
+			sharedCancel()
+			wg.Done()
+		}()
+
+		err := startFunc(cliCtx)
+		if err != nil {
+			logger.Error("halting on error", "err", err)
 		}
-	}()
 
-	// Run the API and return its error, if any
-	return runApi(ctx)
+		errCh <- err
+	}
+
+	go run(runIndexer)
+	go run(runApi)
+
+	err := <-errCh
+
+	wg.Wait()
+
+	return err
 }
```

Comment:
Hm, actually this is implementation tricky because both `runIndexer` and `runApi` encapsulate the `context` object. Canceling `sharedCancel` in this scenario doesn't do anything since the context it's associated with is thrown away. The suggestion of modelling it again `indexer.Run` wasn't a good one, my b.

Lets leave it as you have before and manually shut it down with Ctl-c and figure this out a bit later